### PR TITLE
Detect deleteAll support in KVAO tests

### PR DESCRIPTION
**Pull request by superkhau**
Clear CacheItem data between tests if connector supports `deleteAll`. Tests fail with 501 if this check is not used.

### CODE REVIEW

File: test/kvao/delete-all.suite.js [hunk 1]
```diff
@@ -11,6 +11,8 @@ module.exports = function(dataSourceFactory, connectorCapabilities) {
     let CacheItem;
     beforeEach(function unpackContext() {
       CacheItem = helpers.givenCacheItem(dataSourceFactory);
+      if (helpers.connectorSupports('deleteAll', dataSourceFactory))
+        return CacheItem.deleteAll();
```
Comment (bajtos) [inline test/kvao/delete-all.suite.js]: Actually, I think we should move `CacheItem.deleteAll` into `givenCacheItem`.

Comment (superkhau) [inline test/kvao/delete-all.suite.js]: moved

**Comment by superkhau** [comment]
Closing in favour of https://github.com/strongloop/loopback-datasource-juggler/pull/1214

File: test/kvao/_helpers.js [hunk 2]
```diff
@@ -3,11 +3,22 @@
...
+  const CacheItem = dataSource.createModel('CacheItem', {
     key: String,
     value: 'any',
   });
+  if ('deleteAll' in dataSource.connector)
+    CacheItem.deleteAll();
+  return CacheItem;
+};
+
+exports.givenAnotherModel = function(dataSourceFactory) {
```
Comment (bajtos) [inline test/kvao/_helpers.js]: I think it's time to create a more generic `givenModel` method.

File: test/kvao/_helpers.js [hunk 3]
```diff
@@ -3,11 +3,22 @@
...
 exports.givenCacheItem = function(dataSourceFactory) {
-  var dataSource = dataSourceFactory();
-  return dataSource.createModel('CacheItem', {
+  const dataSource = dataSourceFactory();
+  const CacheItem = dataSource.createModel('CacheItem', {
     key: String,
     value: 'any',
   });
+  if ('deleteAll' in dataSource.connector)
+    CacheItem.deleteAll();
```
Comment (bajtos) [inline test/kvao/_helpers.js]: You need to wait until `deleteAll` returns before starting the tests.

I think we need to change `givenCacheItem` to return a promise now.

```js
exports.givenCacheItem = function(dataSourceFactory) {
  // note - most of this code should be moved to `givenModel`
  // ...
  const p = 'deleteAll' in dataSource.connector ? 
    CacheItem.deleteAll() : Promise.resolve;

    return p.then(() => CacheItem);
};
```

File: test/kvao/delete.suite.js [hunk 4]
```diff
@@ -11,7 +11,6 @@ module.exports = function(dataSourceFactory, connectorCapabilities) {
     let CacheItem;
     beforeEach(function unpackContext() {
       CacheItem = helpers.givenCacheItem(dataSourceFactory);
-      return CacheItem.deleteAll();
```
Comment (bajtos) [inline test/kvao/delete.suite.js]:
```js
beforeEach(function unpackContext() {
  return helpers.givenCacheItem(dataSourceFactory)
    .then(ModelCtor => CacheItem = ModelCtor);
});

File: test/kvao/delete-all.suite.js [hunk 5]
```diff
@@ -23,12 +23,13 @@ module.exports = function(dataSourceFactory, connectorCapabilities) {
     });
 
     it('does not remove data from other existing models', function() {
-      var AnotherModel = dataSourceFactory().createModel('AnotherModel');
+      var AnotherModel = helpers.givenAnotherModel(dataSourceFactory);
```
Comment (bajtos) [inline test/kvao/delete-all.suite.js]:
```js
helpers.givenModel(dataSourceFactory, 'AnotherModel')
  .then(ModelCtor => {
    AnotherModel = ModelCtor;
    return helpers.givenKeys(CacheItem, ['key1', 'key2']);
  })
  .then(() => helpers.givenKeys(AnotherModel, ['key3', 'key4']))
  // ...
```
  
            
  Write

File: test/kvao/delete-all.suite.js [hunk 6]
```diff
@@ -9,11 +9,13 @@ module.exports = function(dataSourceFactory, connectorCapabilities) {
 
   bdd.describeIf(supportsDeleteAll, 'deleteAll', function() {
     let CacheItem;
-    beforeEach(function unpackContext() {
-      CacheItem = helpers.givenCacheItem(dataSourceFactory);
+    beforeEach(function setupCacheItem() {
+      helpers.givenCacheItem(dataSourceFactory)
```
Comment (bajtos) [inline test/kvao/delete-all.suite.js]: missing `return`

**Comment by superkhau** [comment]
@bajtos PTAL, I assume this is good now? CI fails are unrelated.

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
Gonna land this as everything is passing and I addressed all your concerns in the previous comments so I can test the KVXS PR.

**Comment by superkhau** [comment]
Follow up issue for oracle at https://github.com/strongloop/loopback-connector-oracle/issues/100

**Pull request merged**